Threshold

SerenaG opened the pull request:
Closes #570
Closes #303
Closes #329

### Updates
Add Thresholds.

For the line chart, the logic applied will highlight the dots that are above all the thresholds, in both directions (vertical and horizontal).
In order to avoid problems related to colors, this behavior is visually represented with an increasing of the stroke width of the desired dots.

Closes #506.

### Demo screenshot or recording
<img width="820" alt="Screenshot 2020-04-09 at 16 25 01" src="https://user-images.githubusercontent.com/5892387/78905726-b935cc00-7a7e-11ea-9b04-1d2e9a3ffdcf.png">

### Review checklist (for reviewers only)
- [ ] Demos all features
- [ ] Documented/annotated
- [ ] Matches UI/UX specs
- [ ] Meets the [code style guide](https://github.com/IBM/carbon-charts/wiki/Code-style-guide)
- [ ] Accessible
- [ ] Mobile first (responsive)
- [ ] RTL support (bidirectional text)
- [ ] Performant (limited bloat)

caesarsol commented on packages/core/src/components/essentials/threshold.ts:
```diff
@@ -0,0 +1,144 @@
+	appendThresholdLabel() {
+		const holder = select(this.services.domUtils.getHolder());
+		const { value, valueFormatter, fillColor, label = "Threshold" } = this.configs;
+		const formattedValue = valueFormatter ? valueFormatter(value) : value.toLocaleString("en");
```
Is there any way to take the default formatter we actually use in the corresponding axis tick labels?

caesarsol commented on packages/core/src/interfaces/axis-scales.ts:
```diff
@@ -32,6 +33,10 @@ export interface AxisOptions {
 	 */
 	title?: string;
+	/**
+	 * thresholds
```
We probably should include some more details on how the thresholds should be used. Maybe the documentation is clear enough just by reading the typings, but an example would probably help?

natashadecoste commented on packages/core/src/styles/components/_threshold.scss:
```diff
@@ -0,0 +1,49 @@
+    stroke-width: 1;
+    stroke-dasharray: 4;
+    cursor: pointer;
+  }
```
```suggestion
    pointer-events: none;
  }
```

when the cursor is directly over the `<line>` you get a weird flicker because its picking that up instead of the `threshold-hoverable-area`

SerenaG replied:
Thanks, good catch 🙂

natashadecoste commented on packages/core/src/components/essentials/threshold.ts:
```diff
@@ -0,0 +1,166 @@
+		const self = this;
+
+		DOMUtils.appendOrSelect(this.threshold, "rect")
+			.on("mouseover", function() {
```
```suggestion
			.on("mouseover mousemove", function() {
```
since the threshold is using the mouse position to determine it's location, I'm wondering if we would want it to follow the mouse until out of the threshold zone? It definitely seems smoother IMO

SerenaG replied:
Thank you, it works better like this

natashadecoste commented on packages/core/src/styles/components/_threshold.scss:
```diff
@@ -0,0 +1,49 @@
+}
+
+.#{$prefix}--#{$charts-prefix}--threshold--label{
+  background-color: $red-50;
+	pointer-events: none;
+	transition: opacity .1s;
+	@include carbon--motion(standard, expressive);
+	display: inline;
+	box-shadow: 0 1px 6px 0 rgba(0, 0, 0, 0.2);
+	position: absolute;
+	word-wrap: break-word;
+	z-index: 1059;
+  font-family: carbon--font-family("sans-condensed");
+
+  color: $text-01;
+  line-height: 16px;
+	font-size: 12px;
+	padding: 4px;
+	min-width: 20px;
+
```
some weird spacing here

natashadecoste commented on packages/core/src/styles/components/_threshold.scss:
```diff
@@ -0,0 +1,49 @@
+	min-width: 20px;
+
+  &.hidden {
+		opacity: 0;
+		transition: opacity .1s;
+		@include carbon--motion(standard, expressive);
```
```suggestion
	opacity: 0;
	transition: opacity .1s;
	@include carbon--motion(standard, expressive);
```

natashadecoste commented on packages/core/src/styles/graphs/_scatter.scss:
```diff
@@ -12,6 +12,10 @@ g.#{$prefix}--#{$charts-prefix}--scatter {
 		stroke-width: 1.5;
 	}
 
+	circle.dot.threshold-anomaly{
```
```suggestion
	circle.dot.threshold-anomaly {
```

natashadecoste commented on packages/core/src/styles/components/_threshold.scss:
```diff
@@ -0,0 +1,49 @@
+  }
+}
+
+.#{$prefix}--#{$charts-prefix}--threshold--label{
```
```suggestion
.#{$prefix}--#{$charts-prefix}--threshold--label {
```

natashadecoste commented on packages/core/src/styles/components/_threshold.scss:
```diff
@@ -0,0 +1,49 @@
+    stroke-width: 2;
+  }
+
+  rect.threshold-hoverable-area{
```
```suggestion
  rect.threshold-hoverable-area {
```

natashadecoste commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -382,4 +382,39 @@ export class CartesianScales extends Service {
+		if (!thresholds) { return null; }
+
+		const scaleRange = this.getRangeScale();
+		const	higherRangeThreshold = thresholds.sort((a, b) => b.value - a.value)[0];
```
```suggestion
		const higherRangeThreshold = thresholds.sort((a, b) => b.value - a.value)[0];
```

natashadecoste commented on packages/core/src/components/essentials/threshold.ts:
```diff
@@ -0,0 +1,166 @@
+		const formattedValue = valueFormatter ? valueFormatter(value) : this.getFormattedValue();
+
+		this.label = DOMUtils.appendOrSelect(holder, `div.${this.thresholdClass}--label.${this.thresholdIdentifierClass}`);
+		this.label.html(`${label}: ${formattedValue}`);
+		this.label.classed("hidden", true);
+		this.label.style("background-color", fillColor);
```
```suggestion
		this.label.html(`${label}: ${formattedValue}`)
				.classed("hidden", true)
				.style("background-color", fillColor);
```

just a suggestion, in most places we tend to use chain syntax where applicable

SerenaG commented:
Hi @natashadecoste, 
I've added more comments on the code we've added for the threshold. 
Let me know if you meant something different or/and if I can improve.

caesarsol commented:
@theiliad ping :)

theiliad commented on packages/core/demo/data/line.ts:
```diff
@@ -77,6 +79,34 @@ export const lineTimeSeriesOptions = {
+			title: "2019 Annual Sales Figures",
+			mapsTo: "date",
+			scaleType: "time",
+			thresholds: [
```
can this potentially cause issues if you define thresholds on both bottom & top axes?

SerenaG replied:
Do you mean, if we have both the bottom and top axes defined and with thresholds?
Something like:
```javascript
export const lineTimeSeriesWithThresholdsOptions = {
	title: "Line with thresholds (time series)",
	axes: {
		bottom: {
			mapsTo: "date",
			scaleType: "time",
			thresholds: [ {value: new Date(2019, 0, 11), label: "Bottom"}]
		},
		top: {
			scaleType: "time",
			thresholds: [
				{value: new Date(2019, 0, 12), label: "Top", fillColor: "white"}
			]
		},
		right: {
			thresholds: [
				{value: 10000, fillColor: "green", label: "Right"}
			]
		},
		left: {
			mapsTo: "value",
			scaleType: "linear",
			thresholds: [
				{value: 55000, label: "Left", fillColor: "orange"},
				{value: 10000, fillColor: "yellow", label: "Left"}
			]
		}
	},
	curve: "curveMonotoneX"
};
```
<img width="403" alt="Screenshot 2020-04-28 at 12 41 38" src="https://user-images.githubusercontent.com/5892387/80478347-ae6baa00-894d-11ea-8046-2beb521a4f5d.png">

The only problem I see is that if you define the same threshold value for the same orientation (right, left in the example above), the first threshold will override the second one defined. 
But it is left to the user's common sense, what do you think?
Do you see any other issues? 

Did I address your question or did you mean something different @theiliad? 



I've found an issue (as you can see from the screenshot) on horizontal thresholds scale value when both axes are present. I'm working on it!

SerenaG replied:
(Fixed! 🙂)
<img width="308" alt="Screenshot 2020-04-28 at 18 02 02" src="https://user-images.githubusercontent.com/5892387/80509969-6151fd00-897a-11ea-8336-312c58a5c4f3.png">

theiliad replied:
🙂

theiliad commented on packages/core/src/components/axes/two-dimensional-axes.ts:
```diff
@@ -111,5 +117,27 @@ export class TwoDimensionalAxes extends Component {
+		const axesOptions = Tools.getProperty(this.model.getOptions(), "axes", axisPosition);
+		const { thresholds } = axesOptions;
+
+		if (thresholds) {
```
can thresholds move around in data updates? will they render, de-render or re-render properly?

SerenaG replied:
They are rendered, de-render, and re-render when `TwoDimensionalAxes` does, in the same way as axes (TwoDimensionalAxes children) do, 
Should they behave differently?

Do you think it would be better to add a `transition` to them?

theiliad replied:
if it's not too distracting I think a transition might make for a smoother experience

theiliad commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -382,4 +382,40 @@ export class CartesianScales extends Service {
+
+	getRangeDominantThreshold(): null | {threshold: ThresholdOptions, scaleValue: number} {
+		const axesOptions = Tools.getProperty(this.model.getOptions(), "axes");
+		const scaleRangePosition = this.getRangeAxisPosition();
```
```suggestion
		const rangeAxisPosition = this.getRangeAxisPosition();
```

theiliad commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -382,4 +382,40 @@ export class CartesianScales extends Service {
+
+		if (!thresholds) { return null; }
+
+		const scaleRange = this.getRangeScale();
```
```suggestion
		const rangeScale = this.getRangeScale();
```

theiliad commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -382,4 +382,40 @@ export class CartesianScales extends Service {
+
+		const scaleRange = this.getRangeScale();
+		// Find the highest threshold for the range
+		const higherRangeThreshold = thresholds.sort((a, b) => b.value - a.value)[0];
```
```suggestion
		const highestThreshold = thresholds.sort((a, b) => b.value - a.value)[0];
```

theiliad commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -382,4 +382,40 @@ export class CartesianScales extends Service {
+		};
+	}
+
+	getRangeDominantThreshold(): null | {threshold: ThresholdOptions, scaleValue: number} {
```
```suggestion
	getHighestRangeThreshold(): null | {threshold: ThresholdOptions, scaleValue: number} {
```

theiliad commented on packages/core/src/services/scales-cartesian.ts:
```diff
@@ -382,4 +382,40 @@ export class CartesianScales extends Service {
 		return scale;
 	}
+
+	getDomainDominantThreshold(): null | {threshold: ThresholdOptions, scaleValue: number} {
```
```suggestion
	getHighestDomainThreshold(): null | {threshold: ThresholdOptions, scaleValue: number} {
```

theiliad commented on packages/core/src/components/graphs/scatter.ts:
```diff
@@ -77,6 +115,7 @@ export class Scatter extends Component {
 
 		selection.raise()
 			.classed("dot", true)
+			.classed("threshold-anomaly", (d, i) => this.isThresholdsAnomaly(d, i))
```
maybe a comment here as to what this means

theiliad commented on packages/core/src/components/graphs/scatter.ts:
```diff
@@ -57,6 +57,44 @@ export class Scatter extends Component {
 		this.addEventListeners();
 	}
 
+	isThresholdsAnomaly(datum: any, index: number) {
```
maybe a comment here as to what this means, and potentially a more readable name

theiliad commented on packages/core/src/components/graphs/scatter.ts:
```diff
@@ -57,6 +57,44 @@ export class Scatter extends Component {
+
+		const { cartesianScales } = this.services;
+
+		// Get higher domain and range thresholds
```
```suggestion
		// Get highest domain and range thresholds
```

theiliad commented on packages/core/src/components/graphs/scatter.ts:
```diff
@@ -57,6 +57,44 @@ export class Scatter extends Component {
+		const [getXValue, getYValue] = Tools.flipDomainAndRangeBasedOnOrientation(
+			(d, i) => cartesianScales.getDomainValue(d, i),
+			(d, i) => cartesianScales.getRangeValue(d, i),
+			cartesianScales.getOrientation()
```
maybe orientation should be stored in a var since we're using it twice here

theiliad commented:
noticing a few things in the UI:

1. in some of the demos the gridlines are missing

![image](https://user-images.githubusercontent.com/14989804/80987477-19aaf580-8e00-11ea-821d-d73eeabdf594.png)

2. in demos where there are threshold lines, the chart gridlines are also showing up as dashed

![image](https://user-images.githubusercontent.com/14989804/80987317-e23c4900-8dff-11ea-8e2c-948f30076785.png)

3. some of the points here have a thicker border, is that intended?

![image](https://user-images.githubusercontent.com/14989804/80987712-6f7f9d80-8e00-11ea-9950-cb0ab0f2fb31.png)

4. when the ruler shows up, the corresponding points aren't highlighting anymore

![image](https://user-images.githubusercontent.com/14989804/80987863-afdf1b80-8e00-11ea-9d66-62ad3ca559fa.png)

theiliad commented on packages/core/src/components/graphs/scatter.ts:
```diff
@@ -57,6 +57,46 @@ export class Scatter extends Component {
 	}
 
+	// A value is an anomaly if is above all defined domain and range thresholds
+	isValueThresholdsAnomaly(datum: any, index: number) {
```
```suggestion
	isDatapointThresholdAnomaly(datum: any, index: number) {
```

SerenaG commented:
Hi @theiliad, 

Points 1 and 4 were solved by merging master into the branch. 
I've fixed the dashed gridlines (point 2),
Yes, (point 3) the behavior is intended: we highlight points with a thicker border if they are an anomaly for all threshold present (both horizontal and vertical).

I've added a transition to the threshold.
